whois-api/src/main/java/net/ripe/db/whois/api/TopLevelFilter.java

```diff
@@ -0,0 +1,50 @@
+package net.ripe.db.whois.api;
+
+import com.google.common.collect.Maps;
+import net.ripe.db.whois.common.domain.CIString;
+import net.ripe.db.whois.common.etree.NestedIntervalMap;
+import net.ripe.db.whois.common.ip.Interval;
+import net.ripe.db.whois.common.ip.IpInterval;
+import net.ripe.db.whois.common.ip.Ipv4Resource;
+import net.ripe.db.whois.common.ip.Ipv6Resource;
+import net.ripe.db.whois.common.rpsl.ObjectType;
+import net.ripe.db.whois.common.rpsl.RpslObject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class TopLevelFilter<T extends Interval<T>> {
+    private static final HashMap<ObjectType, TopLevelFilter> resourceFilterMap;
+    static {
+        resourceFilterMap = Maps.newHashMap();
+        resourceFilterMap.put(ObjectType.INETNUM, new TopLevelFilter(Ipv4Resource.parse("0/0")));
+        resourceFilterMap.put(ObjectType.INET6NUM, new TopLevelFilter(Ipv6Resource.parse("::/0")));
+    }
+
+
+    private final T root;
+
+    public TopLevelFilter(final T root) {
+        this.root = root;
+    }
+
+    private List<RpslObject> filter(final List<RpslObject> rpslObjects) {
+        final NestedIntervalMap<T, CIString> tree = new NestedIntervalMap<>();
+
+        for (RpslObject rpslObject : rpslObjects) {
+            final T key = (T) IpInterval.parse(rpslObject.getKey());
+            tree.put(key, rpslObject.getKey());
```

Comment:
I suggest only putting top-level objects in the tree - for each rpslObject, check if there is already a less-specific object in the tree, and if so don't add it. Then return the entire tree.